File: 16/solution.java
import java.math.BigInteger;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class solution {
  
  private static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

  private static BigInteger[] reg = new BigInteger[4];


  static private void oper (int op, int a, int b, int c) {
    if       (op ==  0) { reg[c] = reg[a];                                                                      }   // setr "00""08"
    else if  (op ==  1) { reg[c] = reg[a].compareTo(reg[b])>0? BigInteger.ONE: BigInteger.ZERO;                 }   // gtrr "01""12"
    else if  (op ==  2) { reg[c] = reg[a].add(reg[b]);                                                          }   // addr "02""00"
    else if  (op ==  3) { reg[c] = reg[a].and(new BigInteger(""+b));                                            }   // bani "03""05"
    else if  (op ==  4) { reg[c] = new BigInteger(""+a);                                                        }   // seti "04""09"
    else if  (op ==  5) { reg[c] = reg[a].or(new BigInteger(""+b));                                             }   // bori "05""07"
    else if  (op ==  6) { reg[c] = reg[a].or(reg[b]);                                                           }   // borr "06""06"
    else if  (op ==  7) { reg[c] = reg[a].equals(reg[b])? BigInteger.ONE: BigInteger.ZERO;                      }   // eqrr "07""15"
    else if  (op ==  8) { reg[c] = reg[a].multiply(reg[b]);                                                     }   // mulr "08""02"
    else if  (op ==  9) { reg[c] = reg[a].equals(new BigInteger(""+b))?  BigInteger.ONE: BigInteger.ZERO;       }   // eqri "09""14"
    else if  (op == 10) { reg[c] = reg[a].compareTo(new BigInteger(""+b))>0?   BigInteger.ONE: BigInteger.ZERO; }   // gtri "10""11"
    else if  (op == 11) { reg[c] = new BigInteger(""+a).equals(reg[b])?  BigInteger.ONE: BigInteger.ZERO;       }   // eqir "11""13"
    else if  (op == 12) { reg[c] = reg[a].and(reg[b]);                                                          }   // banr "12""04"
    else if  (op == 13) { reg[c] = new BigInteger(""+a).compareTo(reg[b])>0?  BigInteger.ONE: BigInteger.ZERO;  }   // gtir "13""10"
    else if  (op == 14) { reg[c] = reg[a].add(new BigInteger(""+b));                                            }   // addi "14""01"
    else if  (op == 15) { reg[c] = reg[a].multiply(new BigInteger(""+b));                                       }   // muli "15""03"

    return;
  }

  public static void main(String[] args) throws Exception {
    reg[0] = BigInteger.ZERO;
    reg[1] = BigInteger.ZERO;
    reg[2] = BigInteger.ZERO;
    reg[3] = BigInteger.ZERO;

    int i = 0;
    while(reader.ready()) {
      System.out.printf("% 5d % 5d % 5d % 5d % 5d", i++, reg[0].bitLength(), reg[1].bitLength(), reg[2].bitLength(), reg[3].bitLength());
      String[] op = reader.readLine().split("\\s+");
      System.out.printf("%s %s %s %s\n", op[0],op[1],op[2],op[3]);
      oper(Integer.parseInt(op[0]),Integer.parseInt(op[1]),Integer.parseInt(op[2]),Integer.parseInt(op[3]));
    }


    System.out.println(reg[0]);
    System.out.println(reg[1]);
    System.out.println(reg[2]);
    System.out.println(reg[3]);
    System.out.println();
  }
}

File: 16/solution.pl
use strict;
use warnings;
use Switch;
use JSON;

sub oper {
    my ($r,$o) = @_;
    #print "oper:\n@$r\n@$o\n";
    switch($o->[0]) {
        case  0 { $r->[$o->[3]] = $r->[$o->[1]] +  $r->[$o->[2]]; }         # addr
        case  1 { $r->[$o->[3]] = $r->[$o->[1]] +       $o->[2];  }         # addi
        case  2 { $r->[$o->[3]] = $r->[$o->[1]] *  $r->[$o->[2]]; }         # mulr
        case  3 { $r->[$o->[3]] = $r->[$o->[1]] *       $o->[2];  }         # muli
        case  4 { $r->[$o->[3]] = $r->[$o->[1]] &  $r->[$o->[2]]; }         # banr
        case  5 { $r->[$o->[3]] = $r->[$o->[1]] &       $o->[2];  }         # bani
        case  6 { $r->[$o->[3]] = $r->[$o->[1]] |  $r->[$o->[2]]; }         # borr
        case  7 { $r->[$o->[3]] = $r->[$o->[1]] |       $o->[2];  }         # bori
        case  8 { $r->[$o->[3]] = $r->[$o->[1]];                  }         # setr
        case  9 { $r->[$o->[3]] =      $o->[1];                   }         # seti
        case 10 { $r->[$o->[3]] =      $o->[1]  >  $r->[$o->[2]]? 1: 0; }   # gtir
        case 11 { $r->[$o->[3]] = $r->[$o->[1]] >       $o->[2]?  1: 0; }   # gtri
        case 12 { $r->[$o->[3]] = $r->[$o->[1]] >  $r->[$o->[2]]? 1: 0; }   # gtrr
        case 13 { $r->[$o->[3]] =      $o->[1]  == $r->[$o->[2]]? 1: 0; }   # eqir
        case 14 { $r->[$o->[3]] = $r->[$o->[1]] ==      $o->[2]?  1: 0; }   # eqri
        case 15 { $r->[$o->[3]] = $r->[$o->[1]] == $r->[$o->[2]]? 1: 0; }   # eqrr
    }
    #print "@$r\n";
    return $r;
}


my %opmap = ();
for (my $i=0; $i<16; $i++) {
    for (my $j=0; $j<16; $j++) {
        $opmap{$i}{$j} = {};
    }
}

my $N = 0;
while (my $line = <STDIN>) {
    chomp($line);
    next if (length $line) == 0;
   
    my @in = $line =~ /(-?\d+)/g;
    $line = <STDIN>; chomp($line);
    my @op = $line =~ /(-?\d+)/g;
    $line = <STDIN>; chomp($line);
    my @ou = $line =~ /(-?\d+)/g;

    #print "@in\n@op\n@ou\n";

    my $pp = $op[0];
    my $n = 0;
    foreach my $i (0..15) {
        @op[0] = $i;
        my @ii = ();
        push @ii , @in;
        my $r = oper(\@ii,\@op);
        my $test = 1;
        for (my $j = 0; $j < 4; $j++) {
            $test &&= $r->[$j] == $ou[$j];
        }
        if ($test) {
            #printf "op $i\n";
            $n++;
        }
        else {
            delete $opmap{$pp}{$i};
        }
    }
    if ($n >= 3) {
        $N++;
        print "$n\n";
    }

}

print to_json \%opmap , {pretty=>1};
#exit;
my $change = 0;
do {
    $change = 0;
    for (my $i = 0; $i < 16; $i++) {
        my @key = keys %{$opmap{$i}};
        if (scalar @key == 1) {
            
            for (my $j = 0; $j < 16; $j++) {
                next if $i == $j;
                if($opmap{$j}{$key[0]}) {
                    print "deleteing $i $j\n";
                    delete $opmap{$j}{$key[0]};
                    $change = 1;
                }
            }
        }
    }
    print to_json \%opmap , {pretty=>1};
} while ($change);
    
    

print "$N\n";

#my $reg = [0,0,0,0];
#my $i = 0;
#while (my $line = <STDIN>) {
#    printf "%d\n" , $i++;
#    chomp($line);
#    my @op = $line =~ /(-?\d+)/g;
#    $op[0] = 
#    #$op[0] = int($op[0]);
#    #print to_json \@op; print "\n";
#
#    my $reg = oper($reg,\@op);
#    #print to_json $reg; print "\n";
#    #print @$reg; print "\n";
#}
#
#print "@$reg"; print "\n";
